### PR TITLE
Addition of Oxford-IIIT Pet as a New Image Classification and Segmentation Dataset

### LINKED ISSUE #140: [Dataset] Add Oxford-IIIT Pet for image segmentation
Please add the following [Oxford-IIIT Pet Dataset](https://www.robots.ox.ac.uk/~vgg/data/pets/) Dataset as new _Image detection or segmentation_ datasets.

### PR DESCRIPTION
**Pull request by DerrickUnleashed**
## This PR adds support for the **Oxford-IIIT Pet** dataset with functionality for both **image classification** and **image segmentation**.
# Key Features

- Supports multiple target types:
  - `category` (37 pet breeds)
  - `binary-category` (cat vs. dog)
  - `segmentation` (pixel-wise trimap masks)
- Consistent with `cifar10_dataset()`
- Includes a comprehensive test suite with dataloader tests.
- Adds detailed roxygen documentation.

Closes #140

### CODE REVIEW

**Review by cregouby — COMMENTED** [review]
**suggestion** **blocking** Instead of one monolitic dataset, could we split into 3 datasets for each task ?
- `oxfordiiitpet_dataset` multiclass classification, may inherit a lot from classification datasets
- `oxfordiiitpet_binary_dataset` : binary classification dataset
- ``oxfordiiitpet_segmentation_dataset` with the segmented output of class 'image_with_segmentation" that can be directly used by "draw_segmentation_mask" (see

**Comment by cregouby** [comment]
>  I wanted to know if I need to focus on the release or on this task right now.

Release 0.7.0 is on hold. In the meantime, you can keep the contributions to it.

**Comment by DerrickUnleashed** [comment]
Since this is a segmentation dataset. I have made it such that it works similar to our detection datasets. Thus, `$y` returns a named list of `masks` and `label`. For this particular dataset, the label would vary depending on the `target_type` parameter. Below is a small demonstration.

```r
oxfordiiitpet <- oxfordiiitpet_segmentation_dataset(transform = transform_to_tensor)
first_item <- oxfordiiitpet[1]
first_item$x  # RGB image tensor of shape (3, H, W)
first_item$y$masks   # Boolean tensor of shape (3, H, W); one-hot masks for pet, background, and outline respectively
first_item$y$label  # Integer label (1–37 or 1–2 depending on target_type)
oxfordiiitpet$classes[first_item$y$label] # Class name of the label
overlay <- draw_segmentation_masks(first_item$x, masks = first_item$y$masks)
tensor_image_browse(overlay)
```

I'm using `transform = transform_to_tensor` since `draw_segmentation_masks()` requires that the image is a `torch_tensor()`.
I'm also returning `$y$masks` as a `torch_bool()` by default since that would be really helpful for the user in terms of calling `draw_segmentation_masks()`
I'm planning to add two more loaders one is `oxfordiiitpet_dataset()` for categorial target_types and `oxfordiiitpet_binary_dataset()` for binary target_types.

**Comment by cregouby** [comment]
> Since this is a segmentation dataset. I have made it such that it works similar to our detection datasets. Thus, `$y` returns a named list of `masks` and `label`. For this particular dataset, the label would vary depending on the `target_type` parameter. Below is a small demonstration.
> 
> ```r
> oxfordiiitpet <- oxfordiiitpet_segmentation_dataset(transform = transform_to_tensor)
> first_item <- oxfordiiitpet[1]
> first_item$x  # RGB image tensor of shape (3, H, W)
> first_item$y$masks   # Boolean tensor of shape (3, H, W); one-hot masks for pet, background, and outline respectively
> first_item$y$label  # Integer label (1–37 or 1–2 depending on target_type)
> oxfordiiitpet$classes[first_item$y$label] # Class name of the label
> overlay <- draw_segmentation_masks(first_item$x, masks = first_item$y$masks)
> tensor_image_browse(overlay)
> ```
> 
> I'm using `transform = transform_to_tensor` since `draw_segmentation_masks()` requires that the image is a `torch_tensor()`. I'm also returning `$y$masks` as a `torch_bool()` by default since that would be really helpful for the user in terms of calling `draw_segmentation_masks()` I have added two more dataset loaders `oxfordiiitpet_dataset()` and `oxfordiiitpet_binary_dataset()` which work as classification datasets.
> 

**praise** This is the right way to do, thanks. 
**improvement** **blocking**  Please note that you can do it even better so that user can have - in your example- a direct
```
 overlay <- draw_segmentation_masks(first_item)
```
you simply need to conform to and mark `item`s as a `image_with_segmentation_mask` class. 
See 
https://github.com/cregouby/torchvision_/blob/041fa5d3def7ca525b3863473ce549ee80950ba0/R/vision_utils.R#L377-L386

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** clean delivery

File: R/dataset-oxfordiiitpet.R [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+#' @export
+oxfordiiitpet_segmentation_dataset <- torch::dataset(
+  name = "oxfordiiitpet segmentation",
+  resources = list(
+    c("https://www.robots.ox.ac.uk/~vgg/data/pets/data/images.tar.gz", "5c4f3ee8e5d25df40f4fd59a7f44e54c"),
+    c("https://www.robots.ox.ac.uk/~vgg/data/pets/data/annotations.tar.gz", "95a8c909bbe2e81eed6a22bccdf3f68f")
+  ),
+  training_file = "trainval.rds",
+  test_file = "test.rds",
+  archive_size = "800 MB",
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:46]: **improvement** could we be closer to the reality : 
```
(~800 MB) will be downloaded 
...
Content type 'application/octet-stream' length 791918971 bytes (755.2 MB)
...
```

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** clean delivery
**todo**

File: R/dataset-oxfordiiitpet.R [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+#' - \code{x}: a H x W x 3 integer array representing an RGB image.
+#' - \code{y$masks}: a boolean tensor of shape (3, H, W), representing the segmentation trimap as one-hot masks.
+#' - \code{y$label}: an integer representing the class label, depending on the \code{target_type}:
+#'   - \code{"category"}: an integer in 1–37 indicating the pet breed.
+#'   - \code{"binary-category"}: 1 for Cat, 2 for Dog.
+#'
+#' @examples
+#' \dontrun{
+#' # Load the Oxford-IIIT Pet dataset with basic tensor transform
+#' oxfordiiitpet <- oxfordiiitpet_segmentation_dataset(transform = transform_to_tensor)
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:22]: Thanks for the to do. I did run and check the code, but since I already had the file installed, it didn't give any errors, but for a new user, that wouldn't be the case, and they would run into this issue. I'll fix this, and I'll look into making sure I don't make this mistake again.

File: R/dataset-oxfordiiitpet.R [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+      return()
+
+    fs::dir_create(self$raw_folder)
+    fs::dir_create(self$processed_folder)
+
+    cli_inform("Downloading {.cls {class(self)[[1]]}}...")
+
+    for (r in self$resources) {
+      url <- r[1]
+      archive <- download_and_cache(url)
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:93]: Thanks for the suggestion. I've updated it to include a `prefix = class(self)[1]` so users can identify on their hard drive.

File: R/dataset-oxfordiiitpet.R [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+      cli_abort("Dataset not found. You can use `download = TRUE` to download it.")
+
+    data_file <- if (train) self$training_file else self$test_file
+    data <- readRDS(file.path(self$processed_folder, data_file))
+    self$image_paths <- data$image_paths
+    self$labels <- data$labels
+    self$class_to_idx <- data$class_to_idx
+    self$classes <- if (self$target_type == "category") names(self$class_to_idx) else c("Cat", "Dog")
+
+    cli_inform("{.cls {class(self)[[1]]}} dataset loaded with {length(self$image_paths)} images across {length(self$classes)} classes.")
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:78]: Ohh... I didn't think along these lines. So, more personalised UI messages for each dataset leads to more translation work. Yes, Thanks for this todo. I'll update it such that they mimic Caltech UI messages.

**Comment by DerrickUnleashed** [comment]
Thanks for the improvement. It would definitely benefit the user to have the example with `draw_segmentation_masks.image_with_segmentation_mask()` . I have updated it accordingly.

File: R/dataset-oxfordiiitpet.R [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+#'
+#' # Retrieve the image tensor, segmentation mask and label
+#' first_item <- oxfordiiitpet[1]
+#' first_item$x  # RGB image tensor of shape (3, H, W)
+#' first_item$y$masks   # (3, H, W) bool tensor: pet, background, outline
+#' first_item$y$label  # Integer label (1–37 or 1–2 depending on target_type)
+#' oxfordiiitpet$classes[first_item$y$label] # Class name of the label
+#'
+#' # Tag with appropriate class for generic dispatch
+#' class(first_item) <- c("image_with_segmentation_mask", class(first_item))
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:35]: **todo** sorry, I wasn't clear enough. It is not intended to be a end-user duty. It must be made at `.getitem` time. So please remove those two lines. And make sure it works in a test

Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R]: Sure, I’ll make the necessary changes to move that logic into `.getitem` and ensure it’s covered by a test.

Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R]: I have kept `"list"` as a fallback class just in case.

File: R/dataset-oxfordiiitpet.R [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+      utils::untar(archive, exdir = self$raw_folder)
+    }
+
+    for (split in c("trainval", "test")) {
+      ann_file <- file.path(self$raw_folder, "annotations", glue::glue("{split}.txt"))
+      lines <- readLines(ann_file)
+      parts <- strsplit(lines, " ")
+      img_ids <- vapply(parts, `[[`, character(1), 1)
+      labels <- vapply(parts, function(x) as.integer(x[2]), integer(1))
+      bin_labels <- vapply(parts, function(x) as.integer(x[3]), integer(1))
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:112]: **improvement** please use performant integrated function to do so : 
```suggestion
      ann <- read.delim(ann_file, sep = " ", col.names = c("img_ids", "label", "bin_label", "other"), header = FALSE)
```

File: R/dataset-oxfordiiitpet.R [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    masks <- torch_tensor(masks)
+    mask1 <- (masks==1)
+    mask2 <- (masks==2)
+    mask3 <- (masks==3)
+    masks <- torch_stack(list(mask1, mask2, mask3))$to(dtype = torch_bool())
+
+    if (self$target_type == "binary-category") {
+      self$classes <- names(self$class_to_idx)[label]
+      if (substr(self$classes, 1, 1) %in% LETTERS) {
+        label <- as.integer(1)
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:161]: **improvement**
```suggestion
        label <- 1L
```

File: R/dataset-oxfordiiitpet.R [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+      } else {
+        label <- as.integer(2)
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:163]:
```suggestion
        label <- 2L
```

File: R/dataset-oxfordiiitpet.R [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+    fs::file_exists(file.path(self$processed_folder, self$training_file)) && fs::file_exists(file.path(self$processed_folder, self$test_file))
+  },
+
+  .getitem = function(index) {
+    x <- jpeg::readJPEG(self$img_path[index])
+
+    label <- self$labels[index]
+
+    seg_name <- basename(self$img_path[index])
+    mask_path <- file.path(self$raw_folder, "annotations", "trimaps", sub("\\.jpg$", ".png", seg_name))
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:150]:
```suggestion
    seg_name <- tools::file_path_sans_ext(self$img_path[index])
    mask_path <- file.path(self$raw_folder, "annotations", "trimaps", seg_name, ".png" )
```

File: R/dataset-oxfordiiitpet.R [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+
+    if (download){
+      cli_inform("Dataset {.cls {class(self)[[1]]}} (~{.emph {self$archive_size}}) will be downloaded and processed if not already available.")
+      self$download()
+    }
+
+    if (!self$check_exists())
+      cli_abort("Dataset not found. You can use `download = TRUE` to download it.")
+
+    data_file <- if (train) self$training_file else self$test_file
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:74]: **suggestion** **style** please use {} and newline for any if and else (use rstudio snippet or use a styler)

File: R/dataset-oxfordiiitpet.R [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+    data <- readRDS(file.path(self$processed_folder, data_file))
+    self$img_path <- data$image_paths
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:76]: **improvement** **style** don't use late rename, this is a code smell.

File: R/dataset-oxfordiiitpet.R [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    label <- self$labels[index]
+
+    seg_name <- basename(self$img_path[index])
+    mask_path <- file.path(self$raw_folder, "annotations", "trimaps", sub("\\.jpg$", ".png", seg_name))
+    masks <- png::readPNG(mask_path) * 255
+    masks <- torch_tensor(masks)
+    mask1 <- (masks==1)
+    mask2 <- (masks==2)
+    mask3 <- (masks==3)
+    masks <- torch_stack(list(mask1, mask2, mask3))$to(dtype = torch_bool())
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:156]: **improvement** you could save a lot of reassigments, and you don't have to recast into boolean a result of a boolean operation : 
```suggestion
    mask_int <- torch_tensor(png::readPNG(mask_path) * 255)
    masks <- torch_stack(list(mask_int == 1, mask_int == 2, mask_int == 3))
```

File: R/dataset-oxfordiiitpet.R [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+
+    if (!is.null(self$transform))
+      x <- self$transform(x)
+
+    if (!is.null(self$target_transform))
+      y <- self$target_transform(y)
+    
+    item = list(x = x, y = y)
+    class(item) <- c("image_with_segmentation_mask", class(item))
+    item
```
Comment (cregouby) [inline R/dataset-oxfordiiitpet.R:181]: **improvement** you can make all in one operation like in https://github.com/mlverse/torchvision/blob/8acd6907e798c30ea947f4a4b55371026d0af6b4/R/dataset-coco.R#L171C1-L178C1

**Review by cregouby — COMMENTED** [review]
**praise** very close to the result.

File: R/dataset-oxfordiiitpet.R (hunk 12 above)
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:156]: Thanks for this improvement. I agree this is much more simpler.

File: R/dataset-oxfordiiitpet.R (hunk 7 above)
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:161]: I'll update the code accordingly.

File: R/dataset-oxfordiiitpet.R (hunk 6 above)
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:112]: Sorry, I remember that this change was suggested in the previous PRs. Will update the code.

File: R/dataset-oxfordiiitpet.R (hunk 9 above)
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:150]: Thanks for this improvement. It does make the code much more readable and simpler.

File: R/dataset-oxfordiiitpet.R (hunk 10 above)
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:74]: Thanks for the suggestion. I'll make the necessary changes.

File: R/dataset-oxfordiiitpet.R (hunk 13 above)
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R:181]: Thanks for the improvement. I have used a `structure()` as suggested.

File: R/dataset-oxfordiiitpet.R (hunk 11 above)
Comment (DerrickUnleashed) [reply R/dataset-oxfordiiitpet.R]: Or do was it meant that I use data instead of self everywhere ?

Comment (cregouby) [reply R/dataset-oxfordiiitpet.R]: yes: `self$img_path <- data$img_path` is better.

> ```r
> if (train) {
>   data_file <- self$training_file
> } else {
>   data_file <- self$test_file
> }
> 
> data <- readRDS(file.path(self$processed_folder, data_file))
> self$img_path <- data$img_path
> self$labels <- data$labels
> self$class_to_idx <- data$class_to_idx
> ```
> 
> I hope this is what was meant by late rename.

Yest this is what I meant. Thanks
